File: android-app/app/src/main/java/app/t4l/domain/PlanningEngine.kt
package app.t4l.domain

data class PlanningTask(
    val id: String,
    val remainingMinutes: Int,
    val value: Int,
    val energy: String,
    val deadlineEpochMs: Long?,
    val splittable: Boolean,
)

data class PlanningWindow(val startsAtEpochMs: Long, val endsAtEpochMs: Long) {
    val durationMinutes: Int get() = ((endsAtEpochMs - startsAtEpochMs) / 60_000).toInt().coerceAtLeast(0)
}

data class Recommendation(val taskId: String, val suggestedMinutes: Int, val score: Double)

object PlanningEngine {
    fun freeWindows(availability: List<PlanningWindow>, occupied: List<PlanningWindow>): List<PlanningWindow> =
        availability.sortedBy { it.startsAtEpochMs }.flatMap { available ->
            val result = mutableListOf<PlanningWindow>()
            var cursor = available.startsAtEpochMs
            occupied.sortedBy { it.startsAtEpochMs }
                .filter { it.endsAtEpochMs > available.startsAtEpochMs && it.startsAtEpochMs < available.endsAtEpochMs }
                .forEach { block ->
                    if (block.startsAtEpochMs > cursor) {
                        result += PlanningWindow(cursor, minOf(block.startsAtEpochMs, available.endsAtEpochMs))
                    }
                    cursor = maxOf(cursor, block.endsAtEpochMs)
                }
            if (cursor < available.endsAtEpochMs) result += PlanningWindow(cursor, available.endsAtEpochMs)
            result
        }.filter { it.durationMinutes > 0 }

    fun recommend(
        tasks: List<PlanningTask>,
        windowMinutes: Int,
        energy: String,
        nowEpochMs: Long,
        planningFactor: Double = 1.0,
    ): List<Recommendation> = tasks.mapNotNull { task ->
        val adjusted = kotlin.math.ceil(task.remainingMinutes * planningFactor.coerceAtLeast(0.01)).toInt().coerceAtLeast(1)
        if (adjusted > windowMinutes && !task.splittable) return@mapNotNull null
        val urgency = task.deadlineEpochMs?.let { (30.0 - (it - nowEpochMs) / 86_400_000.0).coerceAtLeast(0.0) } ?: 0.0
        val score = task.value + urgency +
            (if (task.energy == energy) 25 else 0) +
            (if (adjusted <= windowMinutes) 10 else 0)
        Recommendation(task.id, minOf(adjusted, windowMinutes), score)
    }.sortedWith(compareByDescending<Recommendation> { it.score }.thenBy { it.suggestedMinutes }.thenBy { it.taskId })

    fun planningFactor(samples: List<Pair<Int, Int>>): Double {
        val ratios = samples.filter { it.first > 0 && it.second >= 0 }.map { it.second.toDouble() / it.first }.sorted()
        if (ratios.size < 3) return 1.0
        val middle = ratios.size / 2
        return if (ratios.size % 2 == 1) ratios[middle] else (ratios[middle - 1] + ratios[middle]) / 2
    }
}

object BudgetEngine {
    fun totalMinutes(categoryId: String, parents: Map<String, String?>, own: Map<String, Int>): Int {
        val visiting = mutableSetOf<String>()
        fun total(id: String): Int {
            require(visiting.add(id)) { "Category cycle detected." }
            val value = (own[id] ?: 0) + parents.filterValues { it == id }.keys.sumOf(::total)
            visiting.remove(id)
            return value
        }
        return total(categoryId)
    }
}

File: server/src/T4L.Api/Domain/PlanningEngine.cs
namespace T4L.Api.Domain;

public sealed record TimeWindow(DateTimeOffset StartsAt, DateTimeOffset EndsAt)
{
    public int DurationMinutes => Math.Max(0, (int)(EndsAt - StartsAt).TotalMinutes);
}

public sealed record TaskRecommendation(Guid TaskId, int SuggestedMinutes, decimal Score);

public static class PlanningEngine
{
    public static IReadOnlyList<TimeWindow> FreeWindows(
        IEnumerable<TimeWindow> availability,
        IEnumerable<TimeWindow> occupied)
    {
        var result = new List<TimeWindow>();
        var blocks = occupied.OrderBy(x => x.StartsAt).ToArray();
        foreach (var available in availability.OrderBy(x => x.StartsAt))
        {
            var cursor = available.StartsAt;
            foreach (var block in blocks.Where(x => x.EndsAt > available.StartsAt && x.StartsAt < available.EndsAt))
            {
                if (block.StartsAt > cursor)
                {
                    result.Add(new TimeWindow(cursor, block.StartsAt < available.EndsAt ? block.StartsAt : available.EndsAt));
                }
                if (block.EndsAt > cursor)
                {
                    cursor = block.EndsAt;
                }
                if (cursor >= available.EndsAt) break;
            }
            if (cursor < available.EndsAt) result.Add(new TimeWindow(cursor, available.EndsAt));
        }
        return result.Where(x => x.DurationMinutes > 0).ToArray();
    }

    public static IReadOnlyList<TaskRecommendation> Recommend(
        IEnumerable<TaskEntity> tasks,
        int windowMinutes,
        EnergyLevel energy,
        DateTimeOffset now,
        decimal planningFactor = 1m)
    {
        if (windowMinutes <= 0) return [];
        if (planningFactor <= 0) planningFactor = 1m;

        return tasks
            .Where(x => x.DeletedAt is null && x.Status == TaskState.Active)
            .Select(task =>
            {
                var adjusted = Math.Max(1, (int)Math.Ceiling(task.RemainingEstimateMinutes * planningFactor));
                var fits = adjusted <= windowMinutes;
                if (!fits && !task.Splittable) return null;
                var suggested = Math.Min(adjusted, windowMinutes);
                var urgency = task.Deadline is null ? 0m : Math.Max(0m, 30m - (decimal)(task.Deadline.Value - now).TotalDays);
                var energyMatch = task.Energy == energy ? 25m : 0m;
                var completionBonus = fits ? 10m : 0m;
                var score = task.Value + urgency + energyMatch + completionBonus;
                return new TaskRecommendation(task.Id, suggested, score);
            })
            .Where(x => x is not null)
            .Cast<TaskRecommendation>()
            .OrderByDescending(x => x.Score)
            .ThenBy(x => x.SuggestedMinutes)
            .ThenBy(x => x.TaskId)
            .ToArray();
    }

    public static decimal PlanningFactor(IEnumerable<(int EstimateMinutes, int ActualMinutes)> samples)
    {
        var ratios = samples
            .Where(x => x.EstimateMinutes > 0 && x.ActualMinutes >= 0)
            .Select(x => (decimal)x.ActualMinutes / x.EstimateMinutes)
            .OrderBy(x => x)
            .ToArray();
        if (ratios.Length < 3) return 1m;
        var middle = ratios.Length / 2;
        return ratios.Length % 2 == 1 ? ratios[middle] : (ratios[middle - 1] + ratios[middle]) / 2m;
    }
}
